sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -1114,10 +1126,20 @@ void Scheduler::GraphBuilder::cleanupFinishedCommands(
     // Collect stream objects for a visited command.
     if (Cmd->getType() == Command::CommandType::RUN_CG) {
       auto ExecCmd = static_cast<ExecCGCommand *>(Cmd);
+
+      // Transfer ownership of stream implementations.
       std::vector<std::shared_ptr<stream_impl>> Streams = ExecCmd->getStreams();
       ExecCmd->clearStreams();
       StreamsToDeallocate.insert(StreamsToDeallocate.end(), Streams.begin(),
                                  Streams.end());
+
+      // Transfer ownership of reduction resources.
+      std::vector<std::shared_ptr<const void>> ReduResources =
```

Comment:
I'm currently working on moving away from finished command cleanup towards cleanup after enqueue in order to address some queue::wait related performance issues, see https://github.com/intel/llvm/pull/5070. Since stream handling is currently tied to post wait traversal, kernels with streams still use finished command cleanup in the linked initial implementation. Host tasks are another exception, for a different reason. Once the implementation is fully complete, `cleanupFinishedCommands` should be dropped entirely.

With this patch CGs with reduction resources would have to be yet another temporary exception waiting for a different solution. So I would really prefer if addressing this didn't involve finished subgraph traversal.

Reply:
Sure. I will convert this to a draft while I think of something.

Reply:
`I'm currently working on moving away from finished command cleanup towards cleanup after enqueue in order to address some queue::wait related performance issues, see https://github.com/intel/llvm/pull/5070`

@sergey-semenov  @steffenlarsen  folks, could you please if this patch is still relevant?

Reply:
With https://github.com/intel/llvm/pull/5070 merged this should be redesigned to work with the new approach. The problem that this PR attempts to solve is still there, but the solution is no longer applicable. I will close the PR but we should still be able to discuss it here.